src/OpenLoco/src/World/StationManager.cpp

```diff
@@ -539,28 +543,93 @@ namespace OpenLoco::StationManager
     // 0x0048F8A0
     StationId allocateNewStation(const World::Pos3 pos, [[maybe_unused]] const CompanyId owner, const uint8_t mode)
     {
-        // Quite a simple function
-        registers regs;
-        regs.ax = pos.x;
-        regs.cx = pos.y;
-        regs.dx = pos.z;
-        regs.bl = mode;
-        // Owner is passed by _updatingCompanyId
-        auto allocated = !(call(0x0048F8A0, regs) & X86_FLAG_CARRY);
-        if (allocated)
+        if (!World::validCoords(pos))
+        {
+            GameCommands::setErrorTitle(StringIds::off_edge_of_map);
+            return StationId::null;
+        }
+
+        // Find first available station
+        for (auto& station : rawStations())
         {
-            return static_cast<StationId>(regs.bx);
+            if (!station.empty())
+                continue;
+
+            auto maybeTown = TownManager::getClosestTownAndDensity(pos);
+            if (!maybeTown)
+            {
+                GameCommands::setErrorTitle(StringIds::town_must_be_built_first);
+                return StationId::null;
+            }
+
+            auto posSmallZ = pos;
+            posSmallZ.z /= World::kSmallZStep;
```

Comment:
This is not required generateNewStationName takes a bigZ.

Reply:
I'm surprised, as there was a `shr` in the z-coord. Could you confirm?